In [61]:
# Importing libraries
import pandas as pd
import numpy as np
import math
import operator
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
import timeit

In [62]:
#### Start of STEP 1
# Importing data 
data = pd.read_csv("iris.csv")
#### End of STEP 1

data.head() 


5.1  3.5  1.4  0.2  Iris-setosa
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa

In [63]:
len(data)

149

# Complexity for n=149, d=4

In [64]:
# Defining a function which calculates euclidean distance between two data points
def euclideanDistance(data1, data2, length):
    distance = 0
    for x in range(length):
        distance += np.square(data1[x] - data2[x])
    return np.sqrt(distance)

def distancesum(x, y, n): 
    sum = 0
      
    # for each point, finding distance 
    # to rest of the point 
    for i in range(n): 
        for j in range(i+1,n): 
            sum += (abs(x[i] - x[j]) +
                        abs(y[i] - y[j])) 
      
    return sum
def cosine(a,b):
    dot = np.dot(a, b)
    norma = np.linalg.norm(a)
    normb = np.linalg.norm(b)
    cos = dot / (norma * normb)
    return cos

# Defining our KNN model
def knn(trainingSet, testInstance, k,m):
 
    distances = {}
    sort = {}
 
    length = testInstance.shape[1]
    
    #### Start of STEP 3
    # Calculating euclidean distance between each row of training data and test data
    for x in range(len(trainingSet)):
        
        #### Start of STEP 3.1
        if m==0:
            dist = euclideanDistance(testInstance, trainingSet.iloc[x], length)
        elif m==1:
            dist = distancesum(testInstance, trainingSet.iloc[x], length)
        else:
            dist = cosine(testInstance, trainingSet.iloc[x])

        distances[x] = dist[0]
        #### End of STEP 3.1
 
    #### Start of STEP 3.2
    # Sorting them on the basis of distance
    sorted_d = sorted(distances.items(), key=operator.itemgetter(1))
    #### End of STEP 3.2
 
    neighbors = []
    
    #### Start of STEP 3.3
    # Extracting top k neighbors
    for x in range(k):
        neighbors.append(sorted_d[x][0])
    #### End of STEP 3.3
    classVotes = {}
    
    #### Start of STEP 3.4
    # Calculating the most freq class in the neighbors
    for x in range(len(neighbors)):
        response = trainingSet.iloc[neighbors[x]][-1]
 
        if response in classVotes:
            classVotes[response] += 1
        else:
            classVotes[response] = 1
    #### End of STEP 3.4

    #### Start of STEP 3.5
    sortedVotes = sorted(classVotes.items(), key=operator.itemgetter(1), reverse=True)
    return(sortedVotes[0][0], neighbors)
    #### End of STEP 3.5

In [65]:
# Creating a dummy testset
testSet = [[7.2, 3.6, 5.1, 2.5]]
test = pd.DataFrame(testSet)

# Time for k=1 & Euclidian Distance

In [66]:
start = timeit.default_timer()
#### Start of STEP 2
# Setting number of neighbors = 1
k = 1
#### End of STEP 2
# Running KNN model
result,neigh = knn(data, test, k,0)

# Predicted class
print(result)
stop = timeit.default_timer()

print('Time: ', stop - start) 

Iris-virginica
Time:  0.17045110700018995


In [67]:
# Nearest neighbor
print(neigh)

[140]


# Time for k=3 & Manhattan Distance

In [69]:
start = timeit.default_timer()
# Setting number of neighbors = 3 
k = 3 
# Running KNN model 
result,neigh = knn(data, test, k,1) 
# Predicted class 
print(result)
stop = timeit.default_timer()

print('Time: ', stop - start) 

Iris-virginica
Time:  0.29375610100032645


In [54]:
# 3 nearest neighbors
print(neigh)

[140, 138, 119]


In [56]:
start = timeit.default_timer()
# Setting number of neighbors = 5
k = 5
# Running KNN model 
result,neigh = knn(data, test, k,1) 
# Predicted class 
print(result)
stop = timeit.default_timer()

print('Time: ', stop - start)

Iris-versicolor
Time:  0.29305921200011653


# Euclidian Distance

In [44]:
x = [ -1, 1, 3, 2 ] 
y = [ 5, 6, 5, 3 ] 
n = len(x) 
print(euclideanDistance(x, y, n) )


8.12403840463596


# Manhattan Distance

In [34]:
x = [ -1, 1, 3, 2 ] 
y = [ 5, 6, 5, 3 ] 
n = len(x) 
print(distancesum(x, y, n) )

22


# Cosine Similarity

In [45]:
x = [ -1, 1, 3, 2 ] 
y = [ 5, 6, 5, 3 ] 
n = len(x) 
print(cosine(x, y) )

0.5827942371086193


In [70]:
#### Start of STEP 1
# Importing data 
data = pd.read_csv("iris2.csv")
#### End of STEP 1

data.head()

1.4  0.2  Iris-setosa
0  1.4  0.2  Iris-setosa
1  1.3  0.2  Iris-setosa
2  1.5  0.2  Iris-setosa
3  1.4  0.2  Iris-setosa
4  1.7  0.4  Iris-setosa

In [71]:
len(data)

107

# Complexity for n=107 and d=2

In [72]:
# Defining a function which calculates euclidean distance between two data points
def euclideanDistance(data1, data2, length):
    distance = 0
    for x in range(length):
        distance += np.square(data1[x] - data2[x])
    return np.sqrt(distance)

def distancesum(x, y, n): 
    sum = 0
      
    # for each point, finding distance 
    # to rest of the point 
    for i in range(n): 
        for j in range(i+1,n): 
            sum += (abs(x[i] - x[j]) +
                        abs(y[i] - y[j])) 
      
    return sum
def cosine(a,b):
    dot = np.dot(a, b)
    norma = np.linalg.norm(a)
    normb = np.linalg.norm(b)
    cos = dot / (norma * normb)
    return cos

# Defining our KNN model
def knn(trainingSet, testInstance, k,m):
 
    distances = {}
    sort = {}
 
    length = testInstance.shape[1]
    
    #### Start of STEP 3
    # Calculating euclidean distance between each row of training data and test data
    for x in range(len(trainingSet)):
        
        #### Start of STEP 3.1
        if m==0:
            dist = euclideanDistance(testInstance, trainingSet.iloc[x], length)
        elif m==1:
            dist = distancesum(testInstance, trainingSet.iloc[x], length)
        else:
            dist = cosine(testInstance, trainingSet.iloc[x])

        distances[x] = dist[0]
        #### End of STEP 3.1
 
    #### Start of STEP 3.2
    # Sorting them on the basis of distance
    sorted_d = sorted(distances.items(), key=operator.itemgetter(1))
    #### End of STEP 3.2
 
    neighbors = []
    
    #### Start of STEP 3.3
    # Extracting top k neighbors
    for x in range(k):
        neighbors.append(sorted_d[x][0])
    #### End of STEP 3.3
    classVotes = {}
    
    #### Start of STEP 3.4
    # Calculating the most freq class in the neighbors
    for x in range(len(neighbors)):
        response = trainingSet.iloc[neighbors[x]][-1]
 
        if response in classVotes:
            classVotes[response] += 1
        else:
            classVotes[response] = 1
    #### End of STEP 3.4

    #### Start of STEP 3.5
    sortedVotes = sorted(classVotes.items(), key=operator.itemgetter(1), reverse=True)
    return(sortedVotes[0][0], neighbors)
    #### End of STEP 3.5

In [73]:
# Creating a dummy testset
testSet = [[5.1, 2.5]]
test = pd.DataFrame(testSet)

# Time for k=1 and Euclidian Distance

In [74]:
start = timeit.default_timer()
#### Start of STEP 2
# Setting number of neighbors = 1
k = 1
#### End of STEP 2
# Running KNN model
result,neigh = knn(data, test, k,0)

# Predicted class
print(result)
stop = timeit.default_timer()

print('Time: ', stop - start) 

Iris-virginica
Time:  0.07508425200012425


# Time for k=3 & Manhattan Distance

In [75]:
start = timeit.default_timer()
# Setting number of neighbors = 3 
k = 3 
# Running KNN model 
result,neigh = knn(data, test, k,1) 
# Predicted class 
print(result)
stop = timeit.default_timer()

print('Time: ', stop - start)

Iris-setosa
Time:  0.05014472500079137


# Time Complexity = O(ndk) 
# for n=149 d=4 and k=1 time=0.17 = t1(say)
# for n=107 d=2 and k=1 time = t1 / ((4/2)*(149/107))= t1/2.78 = 0.17/2.78=0.06
# Observed time = 0.07